### DISCUSSION TITLE
Joe's Friction Log (Archived)

### DISCUSSION THREAD

**Discussion by Mearman** [breadboard-ai/breadboard#1173] [Friction Log]
Thoughts, notes and ponderings

**Comment by Mearman** [comment]
While working on OpenAlex boards, I wanted to reuse the `spread` `code` node.
- First, I tried refactoring it into a separate file and using a TS import to pull it in.
- Then I tried serializing that node and calling it with invoke.
- My most recent attempt was to wrap the `spread` `code` node in its own board and using it via an `invoke`, though that didn't work either. 


related:
- https://github.com/breadboard-ai/breadboard/issues/1160
- https://github.com/breadboard-ai/breadboard/issues/1149

**Reply by Mearman** [reply]
@dglazkov, are any of these methods the "correct" one? Could you point me to a working example?

**Comment by Mearman** [comment]
noticed an issue with controlling flow of the board when a node accepts both known and unknown inputs. 
- see: https://github.com/breadboard-ai/breadboard/issues/1166

**Comment by Mearman** [comment]
came across `.as({})`, not entirely sure of the purpose?
https://github.com/breadboard-ai/breadboard/blob/c4f887be97951d127daccf13ba0acff0a1625ee5/packages/breadboard-web/src/boards/accumulating-context.ts#L43

**Reply by Mearman** [reply]
@dglazkov I'm still curious about this one?

**Reply by dglazkov** [reply]
This one is the "control" wire. It passes no data over it, just control from one node to another.
In the old syntax, it was the "->".

**Reply by Mearman** [reply]
Excellent, that was what I surmised. I need to play around with this more explicitly in my looping boards.

**Comment by Mearman** [comment]
- discovered https://github.com/breadboard-ai/breadboard/issues/1164
- related to:
  - https://github.com/breadboard-ai/breadboard/issues/1162
  - https://github.com/breadboard-ai/breadboard/issues/1163

**Comment by Mearman** [comment]
- related to https://github.com/breadboard-ai/breadboard/issues/1217
In terms of generics, this is a minimal example of a function that would make full use of polymorphic node types
```typescript
function doSomething<T extends Record<string, I>, I>({
  inputKey = "input",
  outputKey = "output",
  ...rest
}: {
  inputKey?: string;
  outputKey?: string;
} & T): Record<string, I> {
  // Ensure inputKey exists in the input object to prevent runtime errors
  if (!(inputKey in rest)) {
    throw new Error(`Input key "${inputKey}" not found in input object.`);
  }

  return {
    [outputKey]: rest[inputKey],
  };
}
```
```typescript
// Example usage:
const result = doSomething({
  input: "value",
  outputKey: "out",
});
console.log(result);
```

**Comment by Mearman** [comment]
Not really friction but in writing the array and object utility nodes, I had a thought that about exporting a `code` node that also wraps the function, so it can be encoded in the board and used with only the core kit. e.g.
```typescript
export function unshift<T>({
  array,
  elements,
}: {
  array: T[];
  elements: T[];
}): {
  array: T[];
} {
  array.unshift(...elements);
  return { array };
}

export const unshiftNodeType: MonomorphicDefinition<
  {
    array: {
      type: AdvancedBreadboardType<unknown[]>;
    };
    elements: {
      type: AdvancedBreadboardType<unknown[]>;
    };
  },
  {
    array: {
      type: AdvancedBreadboardType<unknown[]>;
    };
  }
> = defineNodeType({
  inputs: {
    array: {
      type: array("unknown"),
    },
    elements: {
      type: array("unknown"),
    },
  },
  outputs: {
    array: {
      type: array("unknown"),
    },
  },
  invoke: unshift,
});
```
also providing this
```typescript
export const unshiftCode: Lambda<
  {
    array: unknown[];
    elements: unknown[];
  },
  {
    array: unknown[];
  }
> = code(({ array, elements }) => unshift({ array, elements }));
```

**Comment by Mearman** [comment]
Something that occurred to me yesterday. it's not always clear when a schema is being used for the `in` or `out` ports of a node

**Reply by dglazkov** [reply]
all ports have schemas, whether inferred or specified. Only input and output node allow specifying their ports using a special `schema` port.

**Reply by dglazkov** [reply]
Does that clarify anything 😆

**Reply by Mearman** [reply]
It does. I think my confusion was because of the various points the types can be specified for nodes that are not inputs or outputs
